### PR TITLE
Fix neutron-ibc-transfer-service fee handling

### PR DESCRIPTION
**Pull request by stiiifff**

### CODE REVIEW

**Review by keyleu — DISMISSED** [review]

**Comment by keyleu** [comment]
[send_request payload]: {"action":"query","chain_id":"localneutron-1","cmd":"bank balances neutron14hj2tavq8fpesdwxxcu44rty3hh90vhujrvcmstl4zr3txmfvw9s5c2epq --node=%RPC% --chain-id=%CHAIN_ID% --output=json"}
thread 'main' panicked at local-interchaintest/examples/ibc_transfer_ntrn_juno_gaia_pfm.rs:283:5:
assertion `left == right` failed
  left: 0
 right: 2000
 
 PFM test is failing, is this related to the other PR?

**Comment by stiiifff** [comment]
@keyleu Fixed and tweaked the behavior as follows:
- If the denom being transferred is NTRN:
  - If the full sender's balance is being transferred, the IBC fee is deducted from the account's balance.
  - Otherwise, verify that the account's balance is sufficient to transfer the requested amount + the fee amount (fail otherwise).
- If the denom being transferred is not NTRN, just ensure that the sender account's balance has enough NTRN to cover for the IBC fees.
/cc @bekauz

**Comment by bekauz** [comment]
I think in case denom being transferred is ntrn, all we need to assert is that the amount is gte than the minimum fee. This is slightly annoying because if this is being done on neutron, we query that fee on demand. The non-ntrn transfer case makes sense to me! will give this a look shortly

File: packages/ibc-utils/src/neutron.rs
Comment (bekauz) [inline packages/ibc-utils/src/neutron.rs:51]: I don't think this deduction is necessary, it gets "taken" by the ibc module automatically. Or is this meant to reserve the fees for an additional transfer in the future?

File: packages/ibc-utils/src/neutron.rs
Comment (bekauz) [inline packages/ibc-utils/src/neutron.rs:85]: I know this is not the part of this PR but are we sure we want to use height-based timeouts? I don't think hardcoding revision number to 2 is right

File: packages/ibc-utils/src/neutron.rs
Comment (Art3miX) [reply packages/ibc-utils/src/neutron.rs:51]: AFAIK it needs to know the amount to send, but it can't send the full amount if the denom is NTRN, so we need to reduce the fee to get the amount we can send.

**Review by Art3miX — APPROVED** [review]

File: contracts/services/neutron-ibc-transfer/src/tests.rs
Comment (keyleu) [inline contracts/services/neutron-ibc-transfer/src/tests.rs:183]: left over comment from somewhere else?

File: contracts/services/neutron-ibc-transfer/src/tests.rs
Comment (keyleu) [inline contracts/services/neutron-ibc-transfer/src/tests.rs:180]: Maybe it's worth moving this to somewhere else so that it can be used for any Neutron specific contracts instead of having to copy it

**Review by keyleu — APPROVED** [review]
LGTM, lmk if you want to merge this first before me merging renames

**Comment by stiiifff** [comment]
> I think in case denom being transferred is ntrn, and we are transferring the full available balance then all we need to assert is that the amount is gte than the minimum fee. It will just get subtracted from the total transfer amount and yield the same result. Not 100% sure about what is the case for non-full balance transfers but I suspect that it would get taken from the account balance that is not being transferred. Would be nice to double check this. The non-ntrn transfer case makes sense to me! will give this a look shortly

I have re-checked and .. when transferring the full amount, if we don't deduct the fees, the transfer fails .. so it looks like the fees are not automatically deducted. Re. the difference between a fixed amount and the full balance:

- Fixed amount: transfer that amount precisely, and ensure that the balance is sufficient to also cover for IBC fees, fail otherwise.
- Full balance: transfer whatever is remaining on the account. Only fail if the balance is below the fee amount.

File: packages/ibc-utils/src/neutron.rs
Comment (stiiifff) [reply packages/ibc-utils/src/neutron.rs]: @bekauz hardcoded value was wrong, my bad. And I agree with your remark .. height-based timeouts are a bad idea to deal with, as they require some understanding about the underlying infrastructure (e.g. client revision number). Time-based timeouts provide a simpler (and more stable) abstraction. I have removed the ability to specify height-based timeouts, let me know what you think.

File: packages/ibc-utils/src/generic.rs
Comment (bekauz) [inline packages/ibc-utils/src/generic.rs:26]: does this need to be expressed in nanos? the default timestamp const is expressed in seconds (600). also feels like passing the values for this arg (`Option<u64>`) would mean passing stuff like `60000000000` to express a minute long timeout. feels like seconds are expressive enough, no?

**Review by bekauz — APPROVED** [review]
just one comment regarding the timeout units

File: packages/ibc-utils/src/generic.rs
Comment (keyleu) [reply packages/ibc-utils/src/generic.rs:26]: missed this one, yeah I think it should def be seconds here

File: contracts/services/neutron-ibc-transfer/src/tests.rs
Comment (stiiifff) [reply contracts/services/neutron-ibc-transfer/src/tests.rs:180]: I suggest to do this later, when the need arises .. I don't want to delay the merge of this PR.

File: packages/ibc-utils/src/neutron.rs
Comment (keyleu) [inline packages/ibc-utils/src/neutron.rs:87]: I think this is incorrect and why an overflow is happening right? Should this be nanos? In the generic transfer you are not converting to nanos

File: packages/ibc-utils/src/neutron.rs
Comment (keyleu) [inline packages/ibc-utils/src/neutron.rs:103]: same as above

File: packages/ibc-utils/src/neutron.rs
Comment (keyleu) [reply packages/ibc-utils/src/neutron.rs:87]: nevermind checked the spec and it seems it's correctly in nanos

**Comment by keyleu** [comment]
looking at the previous contract code and this one.

You're doing in contract.rs 
`rust
let ibc_send_msg = valence_ibc_utils::generic::ibc_send_message(
                    env,
                    cfg.remote_chain_info().channel_id.clone(),
                    cfg.output_addr().to_string(),
                    cfg.denom().to_string(),
                    amount.u128(),
                    cfg.memo().clone(),
                    cfg.remote_chain_info()
                        .ibc_transfer_timeout
                        .map(|timeout| block_time.plus_seconds(timeout.u64()).nanos()),
                    cfg.denom_to_pfm_map().clone(),
                )
.map_err(|err| ServiceError::ExecutionError(err.to_string()))?;`

but then again in the ibc_send_message function you are adding the block time plus the timestamp again.

So you are adding it twice and when you are passing it  to the function since you are passing nanos you are overflowing u64

**Review by keyleu — CHANGES_REQUESTED** [review]
needs to fix the  double timestamp addition and use nanos on both generic/neutron version

**Review by keyleu — APPROVED** [review]

**Pull request merged**
